sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CertificateValidationCheck.cs

```diff
@@ -0,0 +1,310 @@
+                    { //For Lambda expression extract location of the parenthesizis only to separate them from secondary location of "true"
+                        var primaryLocation = ((argument.Expression is ParenthesizedLambdaExpressionSyntax Lambda) ? (SyntaxNode)Lambda.ParameterList : argument).GetLocation();
+                        TryReportLocations(new InspectionContext(c), primaryLocation, argument.Expression);
+                    }
+                }
+            }
+        }
+
+        private void TryReportLocations(InspectionContext c, Location primaryLocation, ExpressionSyntax expression)
+        {
+            var locations = ArgumentLocations(c, expression);
+            if (locations.Length != 0)
+            {   //Report both, assignemnt as well as all implementation occurances
+                c.Context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, primaryLocation, additionalLocations: locations));
+            }
+        }
+
+        private static bool IsValidationDelegateType(ITypeSymbol type)
+        {
+            if (type.Is(KnownType.System_Net_Security_RemoteCertificateValidationCallback))
+            {
+                return true;
+            }
+            else if (type is INamedTypeSymbol NamedSymbol && type.Is(KnownType.System_Func_T1_T2_T3_T4_TResult))
+            {
+                //HttpClientHandler.ServerCertificateCustomValidationCallback uses Func<HttpRequestMessage, X509Certificate2, X509Chain, SslPolicyErrors, bool>
+                //We're actually looking for Func<Any Sender, X509Certificate2, X509Chain, SslPolicyErrors, bool>
+                var parameters = NamedSymbol.DelegateInvokeMethod.Parameters;
+                return parameters.Length == 4   //And it should! T1, T2, T3, T4
+                    && parameters[0].Type.IsClassOrStruct() //We don't care about common (Object) nor specific (HttpRequestMessage) type of Sender
+                    && parameters[1].Type.Is(KnownType.System_Security_Cryptography_X509Certificates_X509Certificate2)
+                    && parameters[2].Type.Is(KnownType.System_Security_Cryptography_X509Certificates_X509Chain)
+                    && parameters[3].Type.Is(KnownType.System_Net_Security_SslPolicyErrors)
+                    && NamedSymbol.DelegateInvokeMethod.ReturnType.Is(KnownType.System_Boolean);
+            }
+            return false;
+        }
+
+        private static ImmutableArray<Location> ArgumentLocations(InspectionContext c, ExpressionSyntax expression)
+        {
+            var ret = ImmutableArray.CreateBuilder<Location>();
+            ISymbol MS;
+            switch (expression)
+            {
+                case IdentifierNameSyntax identifier:
+                    MS = c.Context.SemanticModel.GetSymbolInfo(identifier).Symbol;
+                    if (MS != null)
+                    {
+                        foreach (var syntax in MS.DeclaringSyntaxReferences.Select(x => x.GetSyntax()))
```

Comment:
I am not sure to understand why this foreach loop is necessary here. Replacing the for-loop with
```
if (MS != null && MS.DeclaringSyntaxReferences.Length == 1)
{
    ret.AddRange(IdentifierLocations(c, MS.DeclaringSyntaxReferences[0].GetSyntax()));
}
```
does not change the unit test results. Is it possible to have a case where we have more than 1 declaring syntax node in our case ?

Reply:
To avoid unexpected false positives, I would actually only consider the cases the we are sure of. 
For instance, taking the xml as example, if we have a rule raising an issue when all child nodes are true, in the case of partial definition at multiple locations:
```
<!-- Partial def 1 -->
<Notes>
    <Note1>true<Note1>
<Notes>

<!-- Partial def 2 -->
<Notes>
    <Note2>false<Note2>
<Notes>
```
Current rule would raise an issue here, as it found locations where all child nodes were true.
So out of precaution, I would only handle the case where we know that there is a single definition location, as we know what to expect then